File: src/utils/release.rs
use std::{
    env::current_dir,
    fs::File,
    io::{Error, ErrorKind},
};

use self_update::{self_replace::self_replace, update::Release};
use serde::Deserialize;

use super::{api::create_client, system::get_architecture};

pub fn is_major_bump(current_version: &str, new_version: &str) -> bool {
    if new_version == "nightly" {
        return false;
    }

    let current_version = current_version.split(".").collect::<Vec<&str>>();
    let new_version = new_version.split(".").collect::<Vec<&str>>();

    if current_version[0] < new_version[0] {
        return true;
    }

    false
}

#[derive(Deserialize, Debug)]
struct GithubRelease {
    tag_name: String,
}

pub fn get_latest_release() -> Result<String, Error> {
    let url = "https://api.github.com/repos/runtipi/runtipi/releases/latest";

    let http_client = create_client()?;
    let response = http_client.get(url).send();

    match response {
        Ok(response) => {
            if response.status().is_success() {
                let latest = response.json::<GithubRelease>();

                return match latest {
                    Ok(latest) => Ok(latest.tag_name[1..].to_string()),
                    Err(e) => {
                        return Err(Error::new(ErrorKind::Other, format!("Failed to parse latest release: {:?}", e)));
                    }
                };
            } else {
                return Err(Error::new(
                    ErrorKind::Other,
                    format!("Failed to fetch latest release. Status code: {}", response.status()),
                ));
            }
        }
        Err(e) => Err(Error::new(ErrorKind::Other, format!("Error sending request: {:?}", e))),
    }
}

pub fn get_all_releases() -> Result<Vec<Release>, Error> {
    let releases = self_update::backends::github::ReleaseList::configure()
        .repo_owner("runtipi")
        .repo_name("cli")
        .build()
        .map_err(|e| Error::new(ErrorKind::Other, format!("Failed to find releases from GitHub: {:?}", e)))?;

    let fetch_result = releases
        .fetch()
        .map_err(|e| Error::new(ErrorKind::Other, format!("Failed to fetch releases: {:?}", e)))?;

    Ok(fetch_result)
}

pub fn download_release_and_self_relace(release: &Release) -> Result<(), Error> {
    let arch = get_architecture().unwrap_or("x86_64".to_string()).to_string();
    let arch = if arch == "arm64" { "aarch64".to_string() } else { "x86_64".to_string() };

    let asset = release.asset_for(arch.as_str(), Some("linux"));
    let asset = match asset {
        Some(asset) => asset,
        None => {
            return Err(Error::new(
                ErrorKind::Other,
                format!("No asset found for {} {} on release {}", arch, "linux", release.version),
            ));
        }
    };

    let current_dir = current_dir().expect("Unable to get current directory");

    let tmp_dir = tempfile::Builder::new().prefix("self_update").tempdir_in(&current_dir)?;
    let tmp_tarball_path = tmp_dir.path().join(&asset.name);
    let tmp_tarball = File::create(&tmp_tarball_path)?;

    self_update::Download::from_url(&asset.download_url)
        .set_header(reqwest::header::ACCEPT, "application/octet-stream".parse().unwrap())
        .download_to(&tmp_tarball)
        .map_err(|e| Error::new(ErrorKind::Other, format!("Failed to download release: {:?}", e)))?;

    std::process::Command::new("tar")
        .arg("-xzf")
        .arg(&tmp_tarball_path)
        .arg("-C")
        .arg(&current_dir)
        .output()?;

    // asset.name with no extension
    let bin_name = asset.name.split(".").collect::<Vec<&str>>()[0];
    let new_executable_path = current_dir.join(&bin_name);

    std::process::Command::new("chmod").arg("+x").arg(&new_executable_path);

    let result = self_replace(&new_executable_path);
    std::fs::remove_file(new_executable_path)?;

    result
}

File: go-cli/internal/utils/release.go
package utils

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"os/exec"
	"path/filepath"
	"runtime"
	"strings"
	"time"

	"github.com/Masterminds/semver/v3"
	"github.com/runtipi/cli/internal/config"
)

type GitHubRelease struct {
	TagName     string `json:"tag_name"`
	Name        string `json:"name"`
	Prerelease  bool   `json:"prerelease"`
	CreatedAt   string `json:"created_at"`
	PublishedAt string `json:"published_at"`
	Assets      []struct {
		Name        string `json:"name"`
		ContentType string `json:"content_type"`
		DownloadURL string `json:"browser_download_url"`
	} `json:"assets"`
}

func GetLatestRelease() (string, error) {
	url := "https://api.github.com/repos/runtipi/runtipi/releases/latest"

	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		return "", fmt.Errorf("failed to create request: %v", err)
	}

	req.Header.Set("User-Agent", "Runtipi-CLI")
	client := &http.Client{Timeout: 10 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		return "", fmt.Errorf("failed to send request: %v", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return "", fmt.Errorf("failed to fetch latest release. Status code: %d", resp.StatusCode)
	}

	var release GitHubRelease
	if err := json.NewDecoder(resp.Body).Decode(&release); err != nil {
		return "", fmt.Errorf("failed to parse latest release: %v", err)
	}

	return release.TagName, nil
}

func IsMajorBump(currentVersion, newVersion string) bool {
	if newVersion == "nightly" {
		return false
	}

	currentVersion = strings.TrimPrefix(currentVersion, "v")
	newVersion = strings.TrimPrefix(newVersion, "v")

	current, err := semver.NewVersion(currentVersion)
	if err != nil {
		return false
	}

	new, err := semver.NewVersion(newVersion)
	if err != nil {
		return false
	}

	return current.Major() < new.Major()
}

func DownloadReleaseAndSelfReplace(version string) error {
	release, err := FindReleaseByVersion(version)
	if err != nil {
		return fmt.Errorf("%v", err)
	}

	arch := runtime.GOARCH
	switch arch {
	case "arm64":
		arch = "aarch64"
	case "amd64":
		arch = "x86_64"
	}

	var assetURL string
	var assetName string
	for _, asset := range release.Assets {
		if strings.Contains(asset.Name, arch) && strings.Contains(asset.Name, runtime.GOOS) {
			assetURL = asset.DownloadURL
			assetName = asset.Name
			break
		}
	}

	if assetURL == "" {
		return fmt.Errorf("no asset found for %s %s on release %s", arch, runtime.GOOS, release.TagName)
	}

	tempDir, err := os.MkdirTemp(config.RootFolder, "self_update")
	if err != nil {
		return fmt.Errorf("failed to create temporary directory: %v", err)
	}
	defer os.RemoveAll(tempDir)

	req, err := http.NewRequest("GET", assetURL, nil)
	if err != nil {
		return fmt.Errorf("failed to create request: %v", err)
	}

	req.Header.Set("User-Agent", "Runtipi-CLI")
	req.Header.Set("Accept", "application/octet-stream")
	client := &http.Client{Timeout: 10 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		return fmt.Errorf("failed to send request: %v", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return fmt.Errorf("failed to download release. Status code: %d", resp.StatusCode)
	}

	tempFilePath := filepath.Join(tempDir, assetName)
	file, err := os.Create(tempFilePath)
	if err != nil {
		return fmt.Errorf("failed to create file: %v", err)
	}

	_, err = io.Copy(file, resp.Body)
	file.Close()
	if err != nil {
		return fmt.Errorf("failed to write file: %v", err)
	}

	cmd := exec.Command("tar", "-xzf", tempFilePath, "-C", config.RootFolder)
	if err := cmd.Run(); err != nil {
		return fmt.Errorf("failed to extract release: %v", err)
	}

	binName := strings.Split(assetName, ".")[0]
	newExecutablePath := filepath.Join(config.RootFolder, binName)

	err = os.Chmod(newExecutablePath, 0755)
	if err != nil {
		return fmt.Errorf("failed to make binary executable: %v", err)
	}

	err = os.Rename(newExecutablePath, filepath.Join(config.RootFolder, "runtipi-cli"))
	if err != nil {
		return fmt.Errorf("could not rename old binary: %w", err)
	}

	return nil
}

func FindReleaseByVersion(version string) (GitHubRelease, error) {
	// https://api.github.com/repos/runtipi/runtipi/releases/tags/v4.0.0
	url := fmt.Sprintf("https://api.github.com/repos/runtipi/runtipi/releases/tags/%s", version)

	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		return GitHubRelease{}, fmt.Errorf("failed to create request: %v", err)
	}

	req.Header.Set("User-Agent", "Runtipi-CLI")
	client := &http.Client{Timeout: 10 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		return GitHubRelease{}, fmt.Errorf("failed to send request: %v", err)
	}

	defer resp.Body.Close()

	if resp.StatusCode == http.StatusOK {
		var release GitHubRelease
		if err := json.NewDecoder(resp.Body).Decode(&release); err != nil {
			return GitHubRelease{}, fmt.Errorf("failed to parse release: %v", err)
		}
		return release, nil
	}

	return GitHubRelease{}, fmt.Errorf("release not found. Did you forget the v prefix? (e.g. v4.0.0 instead of 4.0.0)")
}
